src/compiler/checker.ts

```diff
@@ -2663,7 +2746,170 @@ namespace ts {
             return links.type;
         }
 
+        function resolveAmdExportAssignment(symbol: Symbol): Type {
+            let seenTypes: Type[] = [];
+
+            for (var i = 0; i < symbol.declarations.length; i++) {
+                let decl = symbol.declarations[i];
+                Debug.assert(isAmdExportAssignment(decl));
+                seenTypes.push(getTypeOfExpression((<BinaryExpression>symbol.declarations[i]).right));
+            }
+
+            return getUnionType(seenTypes);
+        }
+
+        function resolveCommonJsModuleExportsAssignment(symbol: Symbol): Type {
+            let seenTypes: Type[] = [];
+
+            for (var i = 0; i < symbol.declarations.length; i++) {
+                let decl = symbol.declarations[i];
+                Debug.assert(isCommonJsExportsAssignment(decl));
+                seenTypes.push(getTypeOfExpression((<BinaryExpression>symbol.declarations[i]).right));

```

Comment:
Use `checkExpression`, not `getTypeOfExpression`. The latter is a helper function for the language service that performs a check that will never be true here. We should probably rename the language service helper.